File: Generator.hs
{-|
Module: Generator
Description: Generates and enumerates lambda terms

This module contains Haskell implementations for methods to
generate and enumerate various fragments of the lambda
calculus
-}
module Generator where

import Data.List ((\\))

data Term = Var Int | Abs Term | App Term Term deriving (Show)

{-|
  Enumerate all the general lambda terms with subterms n and free variables k.
-}
enumerateTerms :: Int -> Int -> Int
enumerateTerms 0 _ = 0
enumerateTerms 1 k = k
enumerateTerms n k = enumerateTerms (n-1) (k+1)
                         + sum [(enumerateTerms n1 k) * (enumerateTerms (n-1-n1) k) | n1 <- [1..n-2]]

{-|
  Enumerate all the planar lambda terms with subterms n and free variables k.
-}
enumeratePlanarTerms :: Int -> Int -> Int
enumeratePlanarTerms 0 _ = 0
enumeratePlanarTerms 1 1 = 1
enumeratePlanarTerms 1 _ = 0
enumeratePlanarTerms n k = enumeratePlanarTerms (n-1) (k+1)
                         + sum [(enumeratePlanarTerms n1 k1) * (enumeratePlanarTerms (n-1-n1) (k-k1)) | n1 <- [1..n-2], k1 <- [0..k]]

{-|
  Enumerate all the linear lambda terms with subterms n and free variables k.
-}
enumerateLinearTerms :: Int -> Int -> Int
enumerateLinearTerms 0 _ = 0
enumerateLinearTerms 1 0 = 0
enumerateLinearTerms 1 _ = 1
enumerateLinearTerms n k = enumerateLinearTerms (n-1) (k+1)
                         + sum [(choose k k1) * (enumerateLinearTerms n1 k1) * (enumerateLinearTerms (n-1-n1) (k-k1)) | n1 <- [1..n-2], k1 <- [0..k]]


{-|
  Generate all of the general lambda terms with subterms n and free variables k.
-}
generateTerms :: Int -> Int -> [Term]
generateTerms 0 _ = []
generateTerms 1 k = [Var x | x <- [0..k-1]]
generateTerms n k = [Abs t | t <- generateTerms (n-1) (k+1)]
                      ++ [App t1 t2 | n1 <- [1..n-2], t1 <- generateTerms n1 k, t2 <- generateTerms (n-1-n1) k]                 

{-|
  Generate all of the planar lambda terms with subterms n and free variables k.
-}
generatePlanarTerms :: Int -> Int -> [Term]
generatePlanarTerms n k = generatePlanarTerms' n (descList (k-1) 0)

generatePlanarTerms' :: Int -> [Int] -> [Term]
generatePlanarTerms' 0 _ = []
generatePlanarTerms' 1 [k] = [Var k]
generatePlanarTerms' 1 _ = []
generatePlanarTerms' n ks = [Abs t | t <- generatePlanarTerms' (n-1) ((map (+1) ks) ++ [0])]
                      ++ [App t1 t2 | n1 <- [1..n-2], i <- [0..(length ks)], t1 <- generatePlanarTerms' n1 (fst (splitAt i ks)), t2 <- generatePlanarTerms' (n-1-n1) (snd (splitAt i ks))]      

{-|
  Generate all of the linear lambda terms with subterms n and free variables k.
-}
generateLinearTerms :: Int -> Int -> [Term]
generateLinearTerms n k = generateLinearTerms' n (descList (k-1) 0)

generateLinearTerms' :: Int -> [Int] -> [Term]
generateLinearTerms' 0 _ = []
generateLinearTerms' 1 [k] = [Var k]
generateLinearTerms' 1 _ = []
generateLinearTerms' n ks = [Abs t | t <- generateLinearTerms' (n-1) ((map (+1) ks) ++ [0])]
                      ++ [App t1 t2 | n1 <- [1..n-2], ks' <- chooseLists ks, t1 <- generateLinearTerms' n1 (ks'), t2 <- generateLinearTerms' (n-1-n1) (ks \\ ks')]      

                      
chooseLists :: [a] -> [[a]]
chooseLists xs = foldr (++) [] [chooseElems xs k | k <- [0..(length xs)]]

chooseElems :: [a] -> Int -> [[a]]
chooseElems xs k = chooseElems' xs k []

chooseElems' :: [a] -> Int -> [[a]] -> [[a]]
chooseElems' _ 0 _ = [[]]
chooseElems' (x : xs) k acc
          | length (x : xs) == k = acc ++ [(x : xs)]
          | otherwise             = chooseElems' xs k (acc ++ (map (\ys -> x : ys) (chooseElems xs (k-1))))

{-
  Pretty print a lambda term
-}                    
prettyPrint :: Term -> IO()
prettyPrint t = do 
                  prettyPrint' t 0
                  putStr "\n"

prettyPrint' :: Term -> Int -> IO()
prettyPrint' (Var x) _ = putStr (show x)
prettyPrint' (Abs t) 0 =      do 
                                putStr "\\ "
                                prettyPrint' t 0
prettyPrint' (Abs t) x =      do
                                putStr "(\\ "
                                prettyPrint' t (x+1)
                                putStr ")"
prettyPrint' (App t1 t2) 0 =  do
                                prettyPrint' t1 1
                                putStr " "
                                prettyPrint' t2 1
prettyPrint' (App t1 t2) x =  do
                                putStr "("
                                prettyPrint' t1 1
                                putStr " "
                                prettyPrint' t2 x
                                putStr ")"

printTermList :: [Term] -> IO()
printTermList [] = return ()
printTermList (x : xs) = do
                            prettyPrint x
                            printTermList xs

{-
  Generate a descending list of integers from and including a given x down to and including a given y
-}
descList :: Int -> Int -> [Int]
descList x y = descList' x y []

descList' :: Int -> Int -> [Int] -> [Int]
descList' x y xs = if x <= y then xs else descList' (x-1) y (xs ++ [x])

{-
  Compute the binomial coefficient for a given n and r
  Adapted from https://stackoverflow.com/a/6806997
-}
choose :: Int -> Int -> Int
choose _ 0 = 1
choose 0 _ = 0
choose n r = choose (n-1) (r-1) * n `div` r

File: generator.js
/**
 * Functions to generate terms from various fragments of the lambda calculus.
 * 
 * @author George Kaye
 */

var varName = 0;

function genVarName(){
    varName += 1;
    return "b" + varName.toString();
}

/**
 * Generate all pure lambda terms with a given number of subterms and free variables.
 * @param {number} n - The number of subterms.
 * @param {number} k - The number of free variables.
 * @return {Object[]} The array of generated terms.
 */
function generateTerms(n, k){

    varName = 0;

    return generateTermsHelper(n, k, 0);

}

/**
 * Generate all pure lambda terms with a given number of subterms and free variables.
 * @param {number} n - The number of subterms.
 * @param {number} k - The number of free variables.
 * @return {Object[]} The array of generated terms.
 */
function generateTermsHelper(n, k, p){

    var terms = [];


    switch(n){
        case 0:
            break;
        case 1:
            for(i = 0; i <= k-1; i++){
                terms[i] = new LambdaVariable(i, "", 0);
            }
            break;
        default:

            var absTerms = generateTermsHelper(n-1, k+1, p);

            for(i = 0; i < absTerms.length; i++){
                absTerms[i] = new LambdaAbstraction(absTerms[i], genVarName());
            }

            var appTerms = [];
            var x = 0;

            for(var m = 1; m <= n-2; m++){
                
                var lhsTerms = generateTermsHelper(m, k, p);
                var rhsTerms = generateTermsHelper(n-1-m, k, p+1);

                for(a = 0; a < lhsTerms.length; a++){
                    for(b = 0; b < rhsTerms.length; b++){
                        //var p = lhsTerms[a].shiftPosition(0);
                        //rhsTerms[b].shiftPosition(1);

                        appTerms[x] = new LambdaApplication(lhsTerms[a], rhsTerms[b]);
                        x++;
                    }
                }
            }

            terms = absTerms.concat(appTerms);
            break;

    }

    return terms;

}

/**
 * Generate all planar lambda terms with a given number of subterms and free variables.
 * @param {number} n - The number of subterms.
 * @param {number} k - The number of free variables.
 * @return {Object[]} The array of generated terms.
 */
function generatePlanarTerms(n, k){
        
    varName = 0;

    var ks = [];

    for(var i = k - 1; i >= 0; i--){
        ks[k-i-1] = i;
    }

    return generatePlanarTermsHelper(n, ks);
}

/**
 * Generate all planar lambda terms with a given number of subterms and free variables.
 * @param {number} n - The number of subterms.
 * @param {number[]} k - The context containing the free variables.
 * @return {Object[]} The array of generated terms.
 */
function generatePlanarTermsHelper(n, k){

    var terms = [];

    switch(n){
        case 0:
            break;
        case 1:
            
            if(k.length === 1){
                terms[0] = new LambdaVariable(k[0], "");
            }

            break;
        default:

            var absTerms = generatePlanarTermsHelper(n-1, k.map(function(e){return e + 1;}).concat(0));

            for(i = 0; i < absTerms.length; i++){
                absTerms[i] = new LambdaAbstraction(absTerms[i], genVarName());
            }

            var appTerms = [];
            var x = 0;

            for(var m = 1; m <= n-2; m++){
                for(var i = 0; i <= k.length; i++){
            
                    var lhsTerms = generatePlanarTermsHelper(m, k.slice(0, i));
                    var rhsTerms = generatePlanarTermsHelper(n-1-m, k.slice(i));

                    for(a = 0; a < lhsTerms.length; a++){
                        for(b = 0; b < rhsTerms.length; b++){
                            appTerms[x] = new LambdaApplication(lhsTerms[a], rhsTerms[b]);
                            x++;
                        }
                    }
                }
            }

            terms = absTerms.concat(appTerms);
            break;

    }

    return terms;

}

/**
 * Generate all linear lambda terms with a given number of subterms and free variables.
 * @param {number} n - The number of subterms.
 * @param {number} k - The number of free variables.
 * @return {Object[]} The array of generated terms.
 */
function generateLinearTerms(n, k){
    
    varName = 0;

    var ks = [];

    for(var i = k - 1; i >= 0; i--){
        ks[k-i-1] = i;
    }

    return generateLinearTermsHelper(n, ks);
}

/**
 * Generate all linear lambda terms with a given number of subterms and free variables.
 * @param {number} n - The number of subterms.
 * @param {number[]} k - The context containing the free variables.
 * @return {Object[]} The array of generated terms.
 */
function generateLinearTermsHelper(n, k){

    var terms = [];

    switch(n){
        case 0:
            break;
        case 1:
            
            if(k.length === 1){
                terms[0] = new LambdaVariable(k[0], "");
            }

            break;
        default:

            var absTerms = generateLinearTermsHelper(n-1, k.map(function(e){return e + 1;}).concat(0));

            for(i = 0; i < absTerms.length; i++){
                absTerms[i] = new LambdaAbstraction(absTerms[i], genVarName());
            }

            var appTerms = [];
            var x = 0;

            var chooses = chooseArrays(k);

            for(var m = 1; m <= n-2; m++){
                for(var i = 0; i < chooses.length; i++){
            
                    var ks1 = chooses[i];
                    var ks2 = k.filter(function(e){return !(ks1.includes(e));});

                    var lhsTerms = generateLinearTermsHelper(m, ks1);
                    var rhsTerms = generateLinearTermsHelper(n-1-m, ks2);

                    for(a = 0; a < lhsTerms.length; a++){
                        for(b = 0; b < rhsTerms.length; b++){
                            appTerms[x] = new LambdaApplication(lhsTerms[a], rhsTerms[b]);
                            x++;
                        }
                    }
                }
            }

            terms = absTerms.concat(appTerms);
            break;
    }

    return terms;

}

/**
 * Get all the different ways an array can have elements selected from it.
 * @param {Object[]} xs - The array to choose elements from.
 * @return {Object[[]]} All the different ways elements can be chosen.
 */
function chooseArrays(xs){

    var arrays = [];

    for(i = 0; i <= xs.length; i++){
        arrays = arrays.concat(chooseElems(xs, i));
    }

    return arrays;

}

/**
 * Get all the different ways a number of elements can be chosen from an array.
 * @param {Object[]} xs   - The array to choose elements from.
 * @param {number}   k    - The number of elements to choose.
 * @return {Object[[]]} All the different ways elements can be chosen.
 */
function chooseElems(xs, k){
    return chooseElemsHelper(xs, k, []);
}

/**
 * Get all the different ways a number of elements can be chosen from an array.
 * @param {Object[]} xs     - The array to choose elements from.
 * @param {number}   k      - The number of elements to choose.
 * @param {Object[]} acc    - An accumulator.
 * @return {Object[[]]} All the different ways elements can be chosen.
 */
function chooseElemsHelper(xs, k, acc){

    if(k === 0){
        return [[]];
    }

    if(xs.length === k){
        return acc.concat([xs]);
    }

    var otherArrays = chooseElems (xs.slice(1), k-1);

    return chooseElemsHelper(xs.slice(1), k, (acc.concat(otherArrays.map(function(e){return [xs[0]].concat(e)}))));

}